instance/models/openedx_instance.py

```diff
@@ -0,0 +1,178 @@
+        Studio full domain name (eg. 'studio.master.sandbox.opencraft.com')
+        """
+        return '{0.studio_sub_domain}.{0.base_domain}'.format(self)
+
+    @property
+    def studio_url(self):
+        """
+        Studio URL
+        """
+        return u'{0.protocol}://{0.studio_domain}/'.format(self)
+
+    def save(self, **kwargs):
+        """
+        Set default values before saving the instance.
+        """
+        # Set default field values from settings - using the `default` field attribute confuses
+        # automatically generated migrations, generating a new one when settings don't match
+        if not self.base_domain:
+            self.base_domain = settings.INSTANCES_BASE_DOMAIN
+        if self.use_ephemeral_databases is None:
+            self.use_ephemeral_databases = settings.INSTANCE_EPHEMERAL_DATABASES
+        super().save(**kwargs)
+
+    @property
+    def appserver_set(self):
+        """
+        Get the set of OpenEdxAppServers owned by this instance.
+        """
+        return self.ref.openedxappserver_set
+
+    def set_appserver_active(self, appserver_id):
+        """
+        Mark the AppServer with the given ID as the active one.
+        """
+        app_server = self.appserver_set.get(pk=appserver_id)  # Make sure the AppServer is owned by this instance
+        self.active_appserver = app_server
+        self.save()
+        self.logger.info('Updating DNS: LMS at %s...', self.domain)
+        gandi.set_dns_record(type='A', name=self.sub_domain, value=self.active_appserver.server.public_ip)
+        self.logger.info('Updating DNS: Studio at %s...', self.studio_domain)
+        gandi.set_dns_record(type='CNAME', name=self.studio_sub_domain, value=self.sub_domain)
+
+    @log_exception
+    def spawn_appserver(self):
+        """
+        Provision a new AppServer. If it completes successfully, mark it as active.
+        """
+        # Provision external databases:
+        if not self.use_ephemeral_databases:
+            # TODO: Use db row-level locking to ensure we don't get any race conditions when creating these DBs.
+            # Use select_for_update(nowait=True) to lock this object's row, then do these steps, then refresh_from_db
+            self.logger.info('Provisioning MySQL database...')
+            self.provision_mysql()
+            self.logger.info('Provisioning MongoDB databases...')
+            self.provision_mongo()
+            self.logger.info('Provisioning Swift container...')
+            self.provision_swift()
+
+        config_fields = [field.name for field in OpenEdXAppConfiguration._meta.fields if field.name not in ('id', )]
+        config_copy = {field_name: getattr(self, field_name) for field_name in config_fields}
+
+        app_server = self.appserver_set.create(
+            # Name for the app server: this should usually generate a unique name (and won't cause any issues if not):
+            name="AppServer {}".format(self.appserver_set.count() + 1),
```

Comment:
@antoviaque @bradenmacdonald I don't have a strong opinion on this, but here are my two cents: If there are a lot of use cases where we need to look at an AppServer in isolation, it could be helpful to try and use a unique `name` for each app server. In that case it would probably be cleaner to have the AppServer generate a `name` for itself. On the other hand, if we are mostly looking at app servers in the context of instances, it would be nice to number them the way it is currently implemented ("Instance foo: App server 1", "Instance foo: App server 2", ... vs "Instance foo: App server 587", "Instance foo: App server 588", ...).